server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java

```diff
@@ -426,6 +427,48 @@ public void testThresholdLimit() {
         );
     }
 
+    public void testShardSizeDiscrepancyWithinIndex() {
+        var discoveryNodesBuilder = DiscoveryNodes.builder();
+        for (int node = 0; node < 3; node++) {
+            discoveryNodesBuilder.add(newNode("node-" + node));
+        }
+
+        var metadataBuilder = Metadata.builder();
+        var routingTableBuilder = RoutingTable.builder();
+
+        addIndex(metadataBuilder, routingTableBuilder, "testindex", Map.of("node-0", 1, "node-1", 1));
+
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(discoveryNodesBuilder)
+            .metadata(metadataBuilder)
+            .routingTable(routingTableBuilder)
+            .build();
+
+        var index = clusterState.routingTable().index("testindex").getIndex();
+
+        // Even if the two shards in this index vary massively in size, we must compute the balancing threshold to be high enough that
+        // we don't make pointless movements. 500GiB of difference is enough to demonstrate the bug.
+
+        var allocationService = createAllocationService(
+            Settings.EMPTY,
+            () -> new ClusterInfo(
+                Map.of(),
+                Map.of(),
+                Map.of(
+                    ClusterInfo.shardIdentifierFromRouting(new ShardId(index, 0), true),
+                    0L,
+                    ClusterInfo.shardIdentifierFromRouting(new ShardId(index, 1), true),
+                    ByteSizeUnit.GB.toBytes(500)
+                ),
+                Map.of(),
+                Map.of(),
+                Map.of()
+            )
+        );
+
+        assertSame(clusterState, reroute(allocationService, clusterState));
```

Comment:
It would be nice to compare `.getRoutingNodes()` instead of the entire cluster state here.
This would be more specific as for what is not expected to change and would also provide a cleaner diff in case of the test failure.

Reply:
It's important that we return the _same_ cluster state here, not just one that's equal, because we want to avoid an unnecessary cluster state publication. Also IMO on failure it's not too hard to see the difference (IntelliJ links to a diff, for instance).